### DISCUSSION TITLE
What is the reason to not return body of ClientResponse after connecction is closed?

### DISCUSSION THREAD

**Discussion by alufrew** [aio-libs/aiohttp#7905] [Q&A] [answered]
Hi.

I am trying to understand why I can't read body via public read() method after connection is closed. 
I want to have universal function which will make a request of any type and will return response:

```
async def make_request(method, url):
    async with aiohttp.ClientSession() as session:
        async with session.request(method=method, url=url) as resp:
            await resp.read()
            return resp 
```

As far as I can see I can get text() and json() data after conenction is closed. But I cannot get bytes using public method read(). I can do it using private _body as in the code below.

```
async def main():
    resp = await make_request('GET', 'https://dummyjson.com/test')
    print(resp.closed) # connection is closed
    print(await resp.text()) # returns text
    print(await resp.json()) # returns json 
    print(resp._body) # returns body
    print(await resp.read()) # error is risen.
```

I have looked into read() method. And it should return body if it's present. But only inside the context of response. Which is odd, as object already has body.

P.S. Context why I do it whis way. I try to create simplistic api client. I can create my response adapter which will get needed data from ClientResponse but it seems to be an overhead when ClientResponse has all needed info and methods.

```
class ApiClient:
  async def api1():
     return await make_request('GET', 'api_1')

   async def api2():
     return await make_request('POST', 'api_2')
     
class BusinessClient:

      def __init__():
           self._api = ApiClient()
           
       async usefull_function1()
              resp = await._api.api1()
              print(resp.json())
      
      async usefull_function2)
              resp = await._api.api1()
              print(resp.text())            
```

**Comment by Dreamsorcerer** [comment, marked as answer]
> P.S. Context why I do it whis way. I try to create simplistic api client. I can create my response adapter which will get needed data from ClientResponse but it seems to be an overhead when ClientResponse has all needed info and methods.

I would suggest you create that adapter. Even if we changed this, I would still recommend not passing around a Response object outside of its context. It is not designed like that, and we'll probably break your code in some way later on.

Also, creating a whole new ClientSession() in every request isn't very efficient and not doing you any favours. If you don't care about performance, then you can just use https://docs.aiohttp.org/en/stable/client_reference.html#basic-api and drop half that code.

**Reply by alufrew** [reply]
Thank you for your reply.  I very much appreciate it. 

> I would suggest you create that adapter. Even if we changed this, I would still recommend not passing around a Response object outside of its context. It is not designed like that, and we'll probably break your code in some way later on.

As for the part about changing current behaviour. If you will want change it - it seams that read() only needs to return _body as a first action if it is not None.  Something like code below. But I may be mistaken and there is some part of the code i didn't understand.

```
 async def read(self) -> bytes:
        """Read response payload."""
         if self._body is not None:
                return self._body
        
        if  self._released:
            raise ClientConnectionError("Connection closed")
            
         try:
                self._body = await self.content.read()
                for trace in self._traces:
                    await trace.send_response_chunk_received(
                        self.method, self.url, self._body
                    )
                
                return self._body
         except BaseException:
                self.close()
                raise
```

As for the part about creating an adapter. I will look into it. Current behaviour of ClientResponse seems to handle my needs well. But indeed there mey be side effects unknown to me.

> Also, creating a whole new ClientSession() in every request isn't very efficient and not doing you any favours. If you don't care about performance, then you can just use https://docs.aiohttp.org/en/stable/client_reference.html#basic-api and drop half that code.

I will look into basic-api example. I didn't find  that example from the get go. A also think that session per request is overhead. In the future I wil look into sessions to increase performance. But currently I don't see simple way to encapsulate session in ApiClient class without handling close somewhere.

**Comment by Dreamsorcerer** [comment]
> But currently I don't see simple way to encapsulate session in ApiClient class without handling close somewhere.

Correct. You could add `__aenter__()`/`__aexit__()` methods and proxy them through. Then ensure you do `async with ApiClient()` or whatever. That's the recommended approach for such clients in general.
